File: Algorithms/C++/Tree/94_Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> values;

        if (root == NULL) {
            return values;
        }

        stack<TreeNode*> s;
        TreeNode* cur = root;

        while (!s.empty() || cur != NULL) {
            while (cur != NULL) {
                s.push(cur);
                cur = cur->left;
            }

            cur = s.top();
            s.pop();
            values.push_back(cur->val);
            cur = cur->right;
        }

        return values;
    }
};

File: Algorithms/Swift/Tree/94_Binary_Tree_Inorder_Traversal.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func inorderTraversal(root: TreeNode?) -> [Int] {
        var values = [Int]()

        if root == nil {
            return values
        }

        var node = root
        var stack = [TreeNode?]()

        while !stack.isEmpty || node != nil {
            while node != nil {
                stack.append(node)
                node = node!.left
            }

            let current = stack.last!
            stack.removeLast()

            values.append(current!.val)

            node = current!.right
        }

        return values
    }
}
